down-the-stack/dk_template/src/lib.rs

```diff
@@ -0,0 +1,432 @@
+    pub radio: ieee802154::Radio<'static>,
+    /// USBD (Universal Serial Bus Device) peripheral
+    pub usbd: USBD,
+    /// POWER (Power Supply) peripheral
+    pub power: POWER,
+    /// USB control endpoint 0
+    pub ep0in: Ep0In,
+    // --- Exercise --- 🔽
+    /// uarte interface
+    pub uarte: Uarte,
+    // --- Exercise --- 🔼 
+}
+
+/// All LEDs on the board
+pub struct Leds {
+    /// LED1: pin P0.13, green LED
+    pub led_1: Led,
+    /// LED2: pin P0.14, green LED
+    pub led_2: Led,
+    /// LED3: pin P0.15, green LED
+    pub led_3: Led,
+    /// LED4: pin P0.16, green LED
+    pub led_4: Led,
+}
+
+/// A single LED
+pub struct Led {
+    inner: Pin<Output<PushPull>>,
+}
+
+impl Led {
+    /// Turns on the LED
+    pub fn on(&mut self) {
+        defmt::trace!(
+            "setting P{}.{} low (LED on)",
+            if self.inner.port() == Port::Port1 {
+                '1'
+            } else {
+                '0'
+            },
+            self.inner.pin()
+        );
+
+        // NOTE this operations returns a `Result` but never returns the `Err` variant
+        let _ = self.inner.set_low();
+    }
+
+    /// Turns off the LED
+    pub fn off(&mut self) {
+        defmt::trace!(
+            "setting P{}.{} high (LED off)",
+            if self.inner.port() == Port::Port1 {
+                '1'
+            } else {
+                '0'
+            },
+            self.inner.pin()
+        );
+
+        // NOTE this operations returns a `Result` but never returns the `Err` variant
+        let _ = self.inner.set_high();
+    }
+
+    /// Returns `true` if the LED is in the OFF state
+    pub fn is_off(&self) -> bool {
+        self.inner.is_set_high() == Ok(true)
+    }
+
+    /// Returns `true` if the LED is in the ON state
+    pub fn is_on(&self) -> bool {
+        !self.is_off()
+    }
+
+    /// Toggles the state (on/off) of the LED
+    pub fn toggle(&mut self) {
+        if self.is_off() {
+            self.on();
+        } else {
+            self.off()
+        }
+    }
+}
+// --- Exercise ---  🔽
+/// All buttons on the board
+pub struct Buttons {
+    /// BUTTON1: pin P0.11, green LED
+    pub b_1: Button,
+    /// BUTTON2: pin P0.12, green LED
+    pub b_2: Button,
+    /// BUTTON3: pin P0.24, green LED
+    pub b_3: Button,
+    /// BUTTON4: pin P0.25, green LED
+    pub b_4: Button,
+}
+
+/// A single button
+pub struct Button {
+    inner: Pin<Input<PullUp>>,
+}
+
+impl Button {
+    /// returns true if button is pushed 
+    pub fn is_pushed(&self) -> bool {
+        self.inner.is_low() == Ok(true)
+    }
+}
+// --- Exercise --- 🔼 
+
+/// A timer for creating blocking delays
+pub struct Timer {
+    inner: hal::Timer<hal::pac::TIMER0, OneShot>,
+}
+
+
+impl Timer {
+    /// Blocks program execution for at least the specified `duration`
+    pub fn wait(&mut self, duration: Duration) {
+        defmt::trace!("blocking for {:?} ...", duration);
+
+        // 1 cycle = 1 microsecond
```

Comment:
We should explain that this is because Timer is hardcoded to 1 MHz.

```suggestion
        // 1 cycle = 1 microsecond because the underlying HAL driver
        // always sets the timer to 1 MHz.
```

Reply:
Although if you use `delay_us` you don't need this comment at all.